src/main/java/software/amazon/awssdk/crt/http/Http2StreamManager.java

```diff
@@ -0,0 +1,282 @@
+    private final CompletableFuture<Void> shutdownComplete = new CompletableFuture<>();
+
+    /**
+     * Factory function for Http2StreamManager instances
+     *
+     * @param options configuration options
+     * @return a new instance of an Http2StreamManager
+     */
+    public static Http2StreamManager create(HttpStreamManagerOptions options) {
+        return new Http2StreamManager(options);
+    }
+
+    private Http2StreamManager(HttpStreamManagerOptions options) {
+        URI uri = options.getUri();
+        if (uri == null) {
+            throw new IllegalArgumentException("URI must not be null");
+        }
+        if (uri.getScheme() == null) {
+            throw new IllegalArgumentException("URI does not have a Scheme");
+        }
+        if (!HTTP.equals(uri.getScheme()) && !HTTPS.equals(uri.getScheme())) {
+            throw new IllegalArgumentException("URI has unknown Scheme");
+        }
+        if (uri.getHost() == null) {
+            throw new IllegalArgumentException("URI does not have a Host name");
+        }
+
+        ClientBootstrap clientBootstrap = options.getClientBootstrap();
+        if (clientBootstrap == null) {
+            throw new IllegalArgumentException("ClientBootstrap must not be null");
+        }
+
+        SocketOptions socketOptions = options.getSocketOptions();
+        if (socketOptions == null) {
+            throw new IllegalArgumentException("SocketOptions must not be null");
+        }
+
+        boolean useTls = HTTPS.equals(uri.getScheme());
+        TlsContext tlsContext = options.getTlsContext();
+        if (useTls && tlsContext == null) {
+            throw new IllegalArgumentException("TlsContext must not be null if https is used");
+        }
+
+
+        int maxConnections = options.getMaxConnections();
+        if (maxConnections <= 0) {
+            throw new IllegalArgumentException("Max Connections must be greater than zero.");
+        }
+        int maxConcurrentStreamsPerConnection = options.getMaxConcurrentStreamsPerConnection();
+        if (maxConcurrentStreamsPerConnection <= 0) {
+            throw new IllegalArgumentException("Max Concurrent Streams Per Connection must be greater than zero.");
+        }
+        int idealConcurrentStreamsPerConnection = options.getIdealConcurrentStreamsPerConnection();
+        if (idealConcurrentStreamsPerConnection <= 0
+                || idealConcurrentStreamsPerConnection > maxConcurrentStreamsPerConnection) {
+            throw new IllegalArgumentException(
+                    "Ideal Concurrent Streams Per Connection must be greater than zero and smaller than max.");
+        }
+
+        int port = options.getPort();
+        if (port == -1) {
+            port = uri.getPort();
+            /* Pick a default port based on the scheme if one wasn't set */
+            if (port == -1) {
+                if (HTTP.equals(uri.getScheme())) {
+                    port = DEFAULT_HTTP_PORT;
```

Comment:
we don't properly do plaintext HTTP/2, we only do it with "prior knowledge", right?

is this error caught further down the pipe in C? will the error be clearly raised to the user?